python/agents/policy-as-code/policy_as_code_agent/simulation.py

```diff
@@ -0,0 +1,137 @@
+import ast
+import datetime
+import json
+import re
+
+
+def validate_code_safety(code: str) -> list:
+    """
+    Analyzes the code using AST to detect potentially unsafe operations.
+    Returns a list of error messages if unsafe patterns are found.
+    """
+    errors = []
+
+    # Exclude listed modules that should not be imported
+    unsafe_modules = {
+        "os",
+        "sys",
+        "subprocess",
+        "shutil",
+        "pickle",
+        "importlib",
+        "socket",
+        "http",
+        "urllib",
+        "requests",
+    }
+
+    # Exclude listed built-in functions
+    unsafe_functions = {"eval", "exec", "open", "compile", "__import__"}
+
+    try:
+        tree = ast.parse(code)
+    except SyntaxError as e:
+        return [f"Syntax Error in generated code: {e}"]
+
+    for node in ast.walk(tree):
+        # Check imports
+        if isinstance(node, ast.Import):
+            for alias in node.names:
+                if alias.name.split(".")[0] in unsafe_modules:
+                    errors.append(
+                        f"Security Violation: Import of restricted module '{alias.name}' is not allowed."
+                    )
+
+        elif isinstance(node, ast.ImportFrom):
+            if node.module and node.module.split(".")[0] in unsafe_modules:
+                errors.append(
+                    f"Security Violation: Import from restricted module '{node.module}' is not allowed."
+                )
+
+        # Check function calls
+        elif isinstance(node, ast.Call):
+            if isinstance(node.func, ast.Name):
+                if node.func.id in unsafe_functions:
+                    errors.append(
+                        f"Security Violation: Usage of restricted function '{node.func.id}' is not allowed."
+                    )
+
+    return errors
+
+
+def run_simulation(policy_code: str, metadata: list, query: str) -> list:
```

Comment:
query is an unused argument. Should we remove it?

Reply:
Removed the unused `query` argument from `run_simulation` and the `threshold` argument from `find_policy_in_memory`.